### PR TITLE
Edit rustc_span documentation

### PR DESCRIPTION
**Pull request by pierwill**
Various changes to the `rustc_span` docs, including the following:

- Additions to top-level docs
- Edits to the source_map module docs
- Edits to documentation for `Span` and `SpanData`
- Added intra-docs links
- Documentation for Levenshtein distances
- Fixed missing punctuation

### CODE REVIEW

File: compiler/rustc_span/src/span_encoding.rs [hunk 1]
```diff
@@ -1,9 +1,3 @@
-// Spans are encoded using 1-bit tag and 2 different encoding formats (one for each tag value).
```
Comment (pierwill) [inline compiler/rustc_span/src/span_encoding.rs:1]: This appeared out-of-date to me, but I might be wrong.

File: compiler/rustc_span/src/lib.rs [hunk 2]
```diff
@@ -1,4 +1,9 @@
-//! The source positions and related helper functions.
+//! Source positions and related helper functions.
+//!
+//! Important concepts in this module include:
+//!
+//! - the *span*, represented by [`SpanData`] and related types; and
+//! - source code as represented by a [`SourceMap`].
```
Comment (jyn514) [inline compiler/rustc_span/src/lib.rs:6]:
```suggestion
//! - the *span*, represented by [`SpanData`] and related types;
//! - source code as represented by a [`SourceMap`]; and
//! - interned strings, represented by [`Symbol`]s, with some common symbols available statically in the [`sym`] module.
//!
//! Unlike most compilers, the span contains not only the position in the source code, but also various other metadata,
//! such as the edition and macro hygiene. This metadata is stored in [`SyntaxContext`] and [`ExpnData`].
```

File: compiler/rustc_span/src/span_encoding.rs [hunk 3]
```diff
@@ -42,13 +36,11 @@ use rustc_data_structures::fx::FxIndexSet;
 /// - `base` is 32 bits in both `Span` and `SpanData`, which means that `base`
 ///   values never cause interning. The number of bits needed for `base`
 ///   depends on the crate size. 32 bits allows up to 4 GiB of code in a crate.
-///   `script-servo` is the largest crate in `rustc-perf`, requiring 26 bits
-///   for some spans.
```
Comment (jyn514) [inline compiler/rustc_span/src/span_encoding.rs:46]: Why did you remove this?

Comment (pierwill) [reply compiler/rustc_span/src/span_encoding.rs:46]: This seemed like it might be out of date. I didn't see a `rustc-perf` in `compiler/`. Perhaps I'm missing the meaning entirely here.

File: compiler/rustc_span/src/lib.rs (hunk 2 above)
Comment (pierwill) [reply compiler/rustc_span/src/lib.rs:6]: Great stuff! Thanks―:pray:.

File: compiler/rustc_span/src/span_encoding.rs (hunk 3 above)
Comment (jyn514) [reply compiler/rustc_span/src/span_encoding.rs:46]: It's out of tree: https://github.com/rust-lang/rustc-perf

Can you add the comment back?

Comment (pierwill) [reply compiler/rustc_span/src/span_encoding.rs:46]: Ah! Will do.

File: compiler/rustc_span/src/span_encoding.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_span/src/span_encoding.rs:1]: This comment does not seem out of date to me afaict.

This is a short introduction to how the struct `Span` works and the reference to the internal thread seems aslo relevant.

Can you elaborate on why you assume this comment is out of date?

Comment (pierwill) [reply compiler/rustc_span/src/span_encoding.rs:1]: It was just a guess, really. Will put this back.

File: compiler/rustc_span/src/lib.rs [hunk 4]
```diff
@@ -124,7 +133,7 @@ pub enum RealFileName {
 
 impl RealFileName {
     /// Returns the path suitable for reading from the file system on the local host.
-    /// Avoid embedding this in build artifacts; see `stable_name` for that.
+    /// Avoid embedding this in build artifacts; see `stable_name()` for that.
```
Comment (theotherphil) [inline compiler/rustc_span/src/lib.rs:136]: Is there a fixed convention for this? The old comment was referring to a function by name, which seems more correct. If there isn’t a well understood convention I could imagine someone else coming along and reverting these changes as part of similar comment tidying exercise.

Comment (lcnr) [reply compiler/rustc_span/src/lib.rs:136]: i don't think that we have a fixed convention for this

> If there isn’t a well understood convention I could imagine someone else coming along and making the opposite changes as part of a similar comment tidying exercise.

yeah :shrug: i mostly just don't mind either way so imo asking people to remove/add changes like this seems slightly too pedantic to me.
